src/read.rs

```diff
@@ -20,27 +20,21 @@ impl<'a, I> ReadHtml<'a, I> where I: Iterator<Item=Event<'a>> {
 }
 
 impl<'a, I> io::Read for ReadHtml<'a, I> where I: Iterator<Item=Event<'a>> {
-    fn read(&mut self, mut buf: &mut [u8]) -> io::Result<usize> {
-        while self.current.len() < buf.len() {
-            if let Some(next) = self.events.next() {
-                try!(write!(&mut self.current, "{}", next));
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        // TODO handle buf.len() == 0
+        if self.current.len() == 0 {
+            if let Some(event) = self.events.next() {
+                // This feels like an unnecessary redirection. Can we do better?
+                // Write directly to buf and left over to self.current?
```

Comment:
Not that I know of, a slice has to be contiguous in memory. That's why I had a `utf8_len` method on `Event` a while back, so it could check that against `buf.len()` and decide whether to write to the buffer directly or store it in the temporary vector if it was too long to fit in the buffer. I gave up on that after I screwed up the initial implementation and got lots of random characters in the output, far too unsafe having a separate method that has to return the same length as what writing it will actually write.